### PR TITLE
[MONO][Marshal] Fix race condition in marshal callback installation

### LINKED ISSUE #74603: Race condition on marshal-ilgen init
### Description

Because of this change: https://github.com/dotnet/runtime/pull/71203, `mono_marshal_ilgen_init` gets called later, after comnponents are initilized. This allows for potential races in initilizing the callback table.

I propose to fix this by making a pointer to `ilgen_marshal_cb`, and initing that pointer atomically (with a CAS) after the table has been constructed.

See comment here: https://github.com/dotnet/runtime/pull/71203#issuecomment-1227590206

### Reproduction Steps

Race condition; not easily reproducible.

### Expected behavior

Runs

### Actual behavior

```
Assertion at /home/uweigand/runtime/src/mono/mono/component/marshal-ilgen.c:63, condition `!ilgen_cb_inited' not met
```
with this call stack:
```
  #14 0x000003ffb327d0ac in mono_assertion_message (file=0x3ffb33865a0 "/home/uweigand/runtime/src/mono/mono/component/marshal-ilgen.c", line=63, condition=0x3ffb33865e0 "!ilgen_cb_inited") at /home/uweigand/runtime/src/mono/mono/eglib/goutput.c:226
  #15 0x000003ffb3321a30 in mono_install_marshal_callbacks_ilgen (cb=0x3ff9a3f16d0) at /home/uweigand/runtime/src/mono/mono/component/marshal-ilgen.c:63
  #16 0x000003ffb33221ce in mono_marshal_ilgen_init () at /home/uweigand/runtime/src/mono/mono/component/marshal-ilgen.c:2858
  #17 0x000003ffb3322106 in get_marshal_cb () at /home/uweigand/runtime/src/mono/mono/component/marshal-ilgen.c:2753
  #18 0x000003ffb3321f24 in mono_emit_marshal_ilgen (m=0x3ff9a3f1bc8, argnum=0, t=0x3ff8ad98b68, spec=0x0, conv_arg=0, conv_arg_type=0x3ff8ad99ed8, action=MARSHAL_ACTION_CONV_IN, lightweigth_cb=0x3ffb33910e8 <marshal_lightweight_cb>) at /home/uweigand/runtime/src/mono/mono/component/marshal-ilgen.c:2810
  #19 0x000003ffb2e54c52 in mono_emit_marshal (m=0x3ff9a3f1bc8, argnum=0, t=0x3ff8ad98b68, spec=0x0, conv_arg=0, conv_arg_type=0x3ff8ad99ed8, action=MARSHAL_ACTION_CONV_IN) at /home/uweigand/runtime/src/mono/mono/metadata/marshal.c:3226
  #20 0x000003ffb2f219aa in emit_native_wrapper_ilgen (image=0x2aa18a8b950, mb=0x3ff7c90eb50, sig=0x3ff8ad99cc8, piinfo=0x3ff8ad98af0, mspecs=0x3ff7c8f47e0, func=0x3ff82d81750 <CompressionNative_DeflateInit2_>, flags=(EMIT_NATIVE_WRAPPER_CHECK_EXCEPTIONS | EMIT_NATIVE_WRAPPER_RUNTIME_MARSHALLING_ENABLED)) at /home/uweigand/runtime/src/mono/mono/metadata/marshal-lightweight.c:937
  #21 0x000003ffb2e5755a in mono_marshal_emit_native_wrapper (image=0x2aa18a8b950, mb=0x3ff7c90eb50, sig=0x3ff8ad99cc8, piinfo=0x3ff8ad98af0, mspecs=0x3ff7c8f47e0, func=0x3ff82d81750 <CompressionNative_DeflateInit2_>, flags=(EMIT_NATIVE_WRAPPER_CHECK_EXCEPTIONS | EMIT_NATIVE_WRAPPER_RUNTIME_MARSHALLING_ENABLED)) at /home/uweigand/runtime/src/mono/mono/metadata/marshal.c:6300
  #22 0x000003ffb2e56b6c in mono_marshal_get_native_wrapper (method=0x3ff8ad98af0, check_exceptions=1, aot=0) at /home/uweigand/runtime/src/mono/mono/metadata/marshal.c:3672
  #23 0x000003ffb30328e8 in compile_special (method=0x3ff8ad98af0, error=0x3ff9a3f2c38) at /home/uweigand/runtime/src/mono/mono/mini/mini-runtime.c:2469
  #24 0x000003ffb30322b0 in mono_jit_compile_method_with_opt (method=0x3ff8ad98af0, opt=374417919, jit_only=0, error=0x3ff9a3f2c38) at /home/uweigand/runtime/src/mono/mono/mini/mini-runtime.c:2672
  #25 0x000003ffb3031c04 in jit_compile_method_with_opt_cb (arg=0x3ff9a3f28e8) at /home/uweigand/runtime/src/mono/mono/mini/mini-runtime.c:2762
  #26 0x000003ffb3029ea8 in jit_compile_method_with_opt (params=0x3ff9a3f28e8) at /home/uweigand/runtime/src/mono/mono/mini/mini-runtime.c:2778
  #27 0x000003ffb3029c7e in mono_jit_compile_method (method=0x3ff8ad98af0, error=0x3ff9a3f2c38) at /home/uweigand/runtime/src/mono/mono/mini/mini-runtime.c:2797
  #28 0x000003ffb31677ba in common_call_trampoline (regs=0x3ff9a3f2db8, code=0x3ffa806234c "\a\340\300\001", m=0x3ff8ad98af0, vt=0x0, vtable_slot=0x0, error=0x3ff9a3f2c38) at /home/uweigand/runtime/src/mono/mono/mini/mini-trampolines.c:618
  #29 0x000003ffb3166b8e in mono_magic_trampoline (regs=0x3ff9a3f2db8, code=0x3ffa806234c "\a\340\300\001", arg=0x3ff8ad98af0, tramp=0x3ffb3de02c8 "\300X") at /home/uweigand/runtime/src/mono/mono/mini/mini-trampolines.c:759
```

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

ppc, but probably possible on other architectures

### Other information

_No response_

### LINKED ISSUE #77090: [Mono] System.Text.RegularExpressions.Tests source generator tests crash in DeflateInit2_
```
=================================================================
	Managed Stacktrace:
=================================================================
	  at <unknown> <0xffffffff>
	  at <unknown> <0xffffffff>
	  at ZLibStreamHandle:DeflateInit2_ <0x00073>
	  at System.IO.Compression.ZLibNative:CreateZLibStreamForDeflate <0x000a3>
	  at System.IO.Compression.Deflater:.ctor <0x00133>
	  at System.IO.Compression.DeflateStream:InitializeDeflater <0x00087>
	  at System.IO.Compression.DeflateStream:.ctor <0x0006b>
	  at System.IO.Compression.DeflateStream:.ctor <0x0004b>
	  at System.Reflection.PortableExecutable.DebugDirectoryBuilder:WriteEmbeddedPortablePdbData <0x00107>
	  at System.Reflection.PortableExecutable.DebugDirectoryBuilder:AddEmbeddedPortablePdbEntry <0x00087>
	  at Microsoft.Cci.PeWriter:WritePeToStream <0x00cdb>
	  at Microsoft.CodeAnalysis.Compilation:SerializePeToStream <0x001b7>
	  at Microsoft.CodeAnalysis.Compilation:SerializeToPeStream <0x006f7>
	  at Microsoft.CodeAnalysis.Compilation:Emit <0x00587>
```

## Build Information
Log: https://dev.azure.com/dnceng-public/public/_build/results?buildId=53859&view=ms.vss-test-web.build-test-results-tab&runId=1102432&paneView=debug&resultId=196512
Build: https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_build/results?buildId=53859
Build error leg or test failing: System.Text.RegularExpressions.Tests.WorkItemExecution
Pull request: https://github.com/dotnet/runtime/pull/77088

```json
{
  "ErrorMessage": "at ZLibStreamHandle:DeflateInit2_",
  "BuildRetry": false
}
```

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[56035](https://dev.azure.com/dnceng-public/public/_build/results?buildId=56035)|dotnet/runtime|[System.Text.RegularExpressions.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=56035&view=ms.vss-test-web.build-test-results-tab&runId=1149346&resultId=196550)|dotnet/runtime#77170|
|[53859](https://dev.azure.com/dnceng-public/public/_build/results?buildId=53859)|dotnet/runtime|[System.Text.RegularExpressions.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=53859&view=ms.vss-test-web.build-test-results-tab&runId=1102432&resultId=196512)|dotnet/runtime#77088|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|0|2|

### PR DESCRIPTION
**Pull request by naricc**
This fixes a race condition on `ilgen_cb_inited`; this flag was used to determine if callbacks were already installed, but had no synchronization around it. Replaced with a pointer to a dynamically allocated buffer and a `cas`.

Fixes: https://github.com/dotnet/runtime/issues/74603
Fixes: https://github.com/dotnet/runtime/issues/77090

### CODE REVIEW

File: src/mono/mono/metadata/marshal-ilgen.c [hunk 1]
```diff
@@ -29,18 +29,21 @@ static void emit_string_free_icall (MonoMethodBuilder *mb, MonoMarshalConv conv)
...
 static void mono_marshal_ilgen_legacy_init (void);
 
-static gboolean ilgen_cb_inited = FALSE;
-static MonoMarshalIlgenCallbacks ilgen_marshal_cb;
+static MonoMarshalIlgenCallbacks* ilgen_marshal_cb = NULL;
 
 void
 mono_install_marshal_callbacks_ilgen (MonoMarshalIlgenCallbacks *cb)
 {
-	g_assert (!ilgen_cb_inited);
```
[the remark this answers is not in the document]
Comment (naricc) [reply src/mono/mono/metadata/marshal-ilgen.c:38]: @jkotas Was reporting what looks like a race condition here: https://github.com/dotnet/runtime/issues/77090. I'm not sure exactly what interleaving can cause this though.

[the remark this answers is not in the document]
Comment (lambdageek) [reply src/mono/mono/metadata/marshal-ilgen.c:38]: @vargaz @naricc since e107bcbe1190b11a0ce91e06454efe079ec85796, `mono_marshal_ilgen_init` _may_ be called by embedders early, but it doesn't _have to be called_.  In the case where the driver doesn't call it early (for example the normal desktop mono driver doesn't call it), `get_marshal_cb` will initialize lazily.  The race is in the lazy case where one thread can see `ilgen_cb_inited == TRUE`, but the callbacks are all still null.

File: src/mono/mono/metadata/marshal-ilgen.c [hunk 2]
```diff
@@ -29,18 +29,21 @@ static void emit_string_free_icall (MonoMethodBuilder *mb, MonoMarshalConv conv)
...
 	g_assert (cb->version == MONO_MARSHAL_CALLBACKS_VERSION);
-	memcpy (&ilgen_marshal_cb, cb, sizeof (MonoMarshalIlgenCallbacks));
-	ilgen_cb_inited = TRUE;
-}
+	MonoMarshalIlgenCallbacks* local_cb = (MonoMarshalIlgenCallbacks*)malloc(sizeof(MonoMarshalIlgenCallbacks));
```
Comment (lambdageek) [inline src/mono/mono/metadata/marshal-ilgen.c:38]: nit: use `g_malloc`/`g_free`

File: src/mono/mono/metadata/marshal-ilgen.c [hunk 3]
```diff
@@ -29,18 +29,21 @@ static void emit_string_free_icall (MonoMethodBuilder *mb, MonoMarshalConv conv)
...
+	memcpy (local_cb, cb, sizeof (MonoMarshalIlgenCallbacks));
 
+	if (mono_atomic_cas_ptr((void**)&ilgen_marshal_cb, local_cb, NULL  ))
```
Comment (lambdageek) [inline src/mono/mono/metadata/marshal-ilgen.c:41]:
```suggestion
	if (mono_atomic_cas_ptr((void**)&ilgen_marshal_cb, local_cb, NULL  ) != NULL)
```

nit: usually the pattern is `if (mono_atomic_cas_ptr (dest, newValue, expectedValue) != expectedValue)`

File: src/mono/mono/metadata/marshal-ilgen.c [hunk 4]
```diff
@@ -29,18 +29,21 @@ static void emit_string_free_icall (MonoMethodBuilder *mb, MonoMarshalConv conv)
...
 void
 mono_install_marshal_callbacks_ilgen (MonoMarshalIlgenCallbacks *cb)
 {
-	g_assert (!ilgen_cb_inited);
 	g_assert (cb->version == MONO_MARSHAL_CALLBACKS_VERSION);
-	memcpy (&ilgen_marshal_cb, cb, sizeof (MonoMarshalIlgenCallbacks));
-	ilgen_cb_inited = TRUE;
-}
+	MonoMarshalIlgenCallbacks* local_cb = (MonoMarshalIlgenCallbacks*)malloc(sizeof(MonoMarshalIlgenCallbacks));
+	memcpy (local_cb, cb, sizeof (MonoMarshalIlgenCallbacks));
```
Comment (lambdageek) [inline src/mono/mono/metadata/marshal-ilgen.c:39]: I think we need a memory barrier between the memcpy and the CAS

```suggestion
	memcpy (local_cb, cb, sizeof (MonoMarshalIlgenCallbacks));
	mono_memory_barrier ();
```

File: src/mono/mono/metadata/marshal.c [hunk 5]
```diff
@@ -6237,31 +6237,37 @@ mono_marshal_emit_native_wrapper (MonoImage *image, MonoMethodBuilder *mb, MonoM
...
 mono_install_marshal_callbacks (MonoMarshalLightweightCallbacks *cb)
 {
-	g_assert (!lightweight_cb_inited);
 	g_assert (cb->version == MONO_MARSHAL_CALLBACKS_VERSION);
-	memcpy (&marshal_lightweight_cb, cb, sizeof (MonoMarshalLightweightCallbacks));
-	lightweight_cb_inited = TRUE;
+	MonoMarshalLightweightCallbacks* local_cb = (MonoMarshalLightweightCallbacks*)malloc(sizeof(MonoMarshalLightweightCallbacks));
+	memcpy (local_cb, cb, sizeof (MonoMarshalLightweightCallbacks));
+
+	if (mono_atomic_cas_ptr((void**)&marshal_lightweight_cb, local_cb, NULL))
```
Comment (lambdageek) [inline src/mono/mono/metadata/marshal.c:6250]: same comments as the other method:
1. use `g_malloc`/`g_free`
2. barrier after memcpy
3. add `!= NULL` to the `if` condition

**Review by lambdageek — DISMISSED** [review]
lgtm. needs barriers. also some style nits

**Comment by lambdageek** [comment]
> Added fix for race condition.

Please use a better summary. something like `[mono][marshal] Fix race condition in callback initialization`

File: src/mono/mono/metadata/marshal-ilgen.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (vargaz) [reply src/mono/mono/metadata/marshal-ilgen.c:38]: Doesn't it get called during runtime startup ?

[the remark this answers is not in the document]
Comment (lambdageek) [reply src/mono/mono/metadata/marshal-ilgen.c:38]: > Doesn't it get called during runtime startup ?

I was surprised, but apparently it doesn't.  Going back to the very first PR it's always been lazy

[the remark this answers is not in the document]
Comment (lambdageek) [reply src/mono/mono/metadata/marshal-ilgen.c:38]: So the problem is that it's really doing 2 separate things:
1. it's picking if you want ilgen or noilgen marshaling (for example XI without interp wants noilgen, but if interp is on, they want ilgen)
2. it's installing the ilgen callbacks.

we can't install the callbacks until the embedder tells us if they want ilgen or noilgen.  Unfortunately there's only one function right now `mono_marshal_ilgen_init` which both signals "we want ilgen" and installs the ilgen callbacks.  To install the noilgen callbacks we wait until the first time that marshaling gets used.

So tl;dr we can't just call `mono_marshal_ilgen_init` from runtime initialization because that might not be what the user wants.

We could probably install the callbacks (either ilgen or noilgen) late during startup, and use `mono_marshal_ilgen_init` just to set a flag that we want ilgen. That's essentially what Nathan's componentization work will do.  but that's actually a riskier change than this.  And we might need to backport to 7 if this race condition shows up there, too.

[the remark this answers is not in the document]
Comment (lambdageek) [reply src/mono/mono/metadata/marshal-ilgen.c:38]: Well I don't have any objection to just using `mono_marshal_ilgen_init` to set a flag and doing all the callback initialization late during startup.  @naricc I think that's what you wanted to do for the component in the first place right?

[the remark this answers is not in the document]
Comment (naricc) [reply src/mono/mono/metadata/marshal-ilgen.c:38]: @lambdageek That is what I am doing for the componentization change.  

So is that what we should do here too, instead of the thing with atomics? `mono_marshal_ilgen_init` sets a flag; we check that flag during, say, `mini_init()`. At which point we do callback installation if it is set?

[the remark this answers is not in the document]
Comment (naricc) [reply src/mono/mono/metadata/marshal-ilgen.c:38]: I guess in the case that we are using NOILGEN and no one has called `mono_marshal_ilgen_init` we just want to go ahead and install the noilgen callbacks?

[the remark this answers is not in the document]
Comment (lambdageek) [reply src/mono/mono/metadata/marshal-ilgen.c:38]: > 

Yea that sounds right.

> So is that what we should do here too, instead of the thing with atomics? mono_marshal_ilgen_init sets a flag; we check that flag during, say, mini_init(). At which point we do callback installation if it is set?

Yea, basically at the point where in the componentized versino we'd call the component init function, here we'd just set up the callbacks.

I think writing down https://github.com/dotnet/runtime/pull/77383#discussion_r1003661360 helped me to think about what the real problem is.   It's not concurrency.  It's that we have a confusing API for choosing between ilgen/noilgen.  We should init the marshaling stack at startup - there's no need for it to be lazy.   You were right in your proposal about `mono_marshal_ilgen_init` (just use it to set a flag), but I didn't understand the problem properly.

**Comment by naricc** [comment]
I am abandoing this PR in favor of doing it the flag way ( see discussion above). That is a different change, so I will do it in a different PR.
